### PR TITLE
Updated Match Patterns

### PR DESCRIPTION
**Pull request by wizardofdocs**
Condensed descriptions and examples. Converted example table to definition list. Integrated the invalid list with the descriptions and examples, to not confuse the reader with a list of things to not do.

Please make sure the special cases and examples are up to date.

Note on the invalid list: It's standard technical writing practice to state instructions positively whenever possible. Readers skimming a document are likely to confuse a list of don'ts with a list of dos and make more mistakes than if the don'ts weren't listed.

### CODE REVIEW

File: site/en/docs/extensions/mv3/match_patterns/index.md
Comment (jpmedley) [inline site/en/docs/extensions/mv3/match_patterns/index.md:16]:
```suggestion
**scheme**: Must be one of `'http'`, `'https'`, `'file'`, `'ftp'`, or `'urn'`. The wildcard `*` only matches `http` or `https`.
```

Comment (jpmedley) [reply site/en/docs/extensions/mv3/match_patterns/index.md:16]: The run-in heading does an adequate job of making this scannable, so that nested bullets aren't needed.

File: site/en/docs/extensions/mv3/match_patterns/index.md
Comment (jpmedley) [inline site/en/docs/extensions/mv3/match_patterns/index.md:19]:
```suggestion
**host**: A hostname (`www.example.com`). A `*` before the hostname to match subdomains (`*.example.com`), or just a wildcard `*`.
```

File: site/en/docs/extensions/mv3/match_patterns/index.md
Comment (jpmedley) [inline site/en/docs/extensions/mv3/match_patterns/index.md:21]: @AmySteam do you know what the `/*` is trying to tell me?

Comment (jpmedley) [inline site/en/docs/extensions/mv3/match_patterns/index.md:21]:
```suggestion
**path**: Must contain at least a forward slash. The slash by itself is treated as a wildcard `/*`. For example:
```

File: site/en/docs/extensions/mv3/match_patterns/index.md
Comment (jpmedley) [inline site/en/docs/extensions/mv3/match_patterns/index.md:25]: This seems like it should be a bulleted list. 

@oliverdunk @AmySteam Are we missing any places where match patterns can be used?

File: site/en/docs/extensions/mv3/match_patterns/index.md
Comment (jpmedley) [inline site/en/docs/extensions/mv3/match_patterns/index.md:29]:
```suggestion
`"<all_urls>"`
```

Comment (jpmedley) [reply site/en/docs/extensions/mv3/match_patterns/index.md:29]: Quotes are part of the syntax of this items.

File: site/en/docs/extensions/mv3/match_patterns/index.md
Comment (jpmedley) [inline site/en/docs/extensions/mv3/match_patterns/index.md:32]:
```suggestion
`"file:///"`
```

Comment (jpmedley) [reply site/en/docs/extensions/mv3/match_patterns/index.md:32]: @oliverdunk @AmySteam This looks to me like it should only have two slashes.

File: site/en/docs/extensions/mv3/match_patterns/index.md
Comment (jpmedley) [inline site/en/docs/extensions/mv3/match_patterns/index.md:33]:
```suggestion
: Allows your extension to run on local files. This pattern requires the user to manually [grant access][permissions-allow-access].
```

Comment (jpmedley) [reply site/en/docs/extensions/mv3/match_patterns/index.md:33]: File URLs never have a host.

**Review by jpmedley — CHANGES_REQUESTED** [review]

File: site/en/docs/extensions/mv3/match_patterns/index.md
Comment (AmySteam) [reply site/en/docs/extensions/mv3/match_patterns/index.md:32]: I've always used three slashes on local files in extensions. This works in my samples.
https://en.wikipedia.org/wiki/File_URI_scheme

**Comment by AmySteam** [comment]
The formatting needs to be tweaked
Screenshot 2023-08-31 at 16 18 32

**Comment by wizardofdocs** [comment]
> The formatting needs to be tweaked Screenshot 2023-08-31 at 16 18 32

Done. Thanks for catching that.

File: site/en/docs/extensions/mv3/match_patterns/index.md
Comment (oliverdunk) [reply site/en/docs/extensions/mv3/match_patterns/index.md:35]: I agree it's slightly weird, but I think three is right. The first two are the protocol (similar to http://) and the last one is for the slash at the start of a file path.

File: site/en/docs/extensions/mv3/match_patterns/index.md
Comment (jpmedley) [inline site/en/docs/extensions/mv3/match_patterns/index.md:36]:
```suggestion
: Allows your extension to run on local files. This pattern requires the user to manually [grant access][permissions-allow-access]. Note that this case requires three slashes, not two.
```

Comment (jpmedley) [reply site/en/docs/extensions/mv3/match_patterns/index.md:36]: This link is broken. Please add this to the list at the bottom:

[permissions-allow-access]: /docs/extensions/mv3/declare_permissions/#allow_access

File: site/en/docs/extensions/mv3/match_patterns/index.md
Comment (jpmedley) [reply site/en/docs/extensions/mv3/match_patterns/index.md:35]: @wizardofdocs I added some text to the definition to explain this.

**Review by jpmedley — CHANGES_REQUESTED** [review]

File: site/en/docs/extensions/mv3/match_patterns/index.md
Comment (oliverdunk) [inline site/en/docs/extensions/mv3/match_patterns/index.md:21]: Is this right? Later in the same article we talk about things like `/foo*`.

File: site/en/docs/extensions/mv3/match_patterns/index.md
Comment (oliverdunk) [reply site/en/docs/extensions/mv3/match_patterns/index.md:28]: I can't think of any others - thanks for checking!

**Review by oliverdunk — CHANGES_REQUESTED** [review]

**Comment by oliverdunk** [comment]
Thanks for the work @wizardofdocs! Really glad we are finding the time for some spring cleaning :)

File: site/en/docs/extensions/mv3/match_patterns/index.md
Comment (jpmedley) [reply site/en/docs/extensions/mv3/match_patterns/index.md:21]: I don't know. @AmySteam?

File: site/en/docs/extensions/mv3/match_patterns/index.md
Comment (wizardofdocs) [reply site/en/docs/extensions/mv3/match_patterns/index.md:28]: To clean up the bullet list structure: what do WAR and Externally Connectable do with the match patterns? It sounds to me like those aren't part of the declaring permissions part of the sentence.

Comment (jpmedley) [reply site/en/docs/extensions/mv3/match_patterns/index.md:28]: The match patterns are for URLs. They effectively grant access permission to URLs.

Comment (wizardofdocs) [reply site/en/docs/extensions/mv3/match_patterns/index.md]: Updated. How's this?

Comment (jpmedley) [reply site/en/docs/extensions/mv3/match_patterns/index.md]: That looks good. Can you make it a bulleted list?

Comment (AmySteam) [reply site/en/docs/extensions/mv3/match_patterns/index.md]: ## Original

So originally I separated the permissions part with the "Other ways that extensions use match patterns".

Screenshot 2023-09-08 at 16 28 28
 
## New approach

The new approach groups it all together as "permissions", which is not accurate.

Screenshot 2023-09-08 at 16 27 37

## Explanation

The match patterns under **web accessible resources** declare which websites can access specific resources that are bundled with the Chrome extension.

The match patterns under "externally_connectable" declare which websites can send and receive messages from the Chrome extension. 

These last two use cases IMO don't have anything to do with permissions.

## What problem are we solving?

I think the original version was clear enough without the need of a rewrite. I'm still confused about what problem we are solving with rewriting this section

Comment (wizardofdocs) [reply site/en/docs/extensions/mv3/match_patterns/index.md]: That was my question.
I was asked to make that long sentence into a bulleted list to make it easier to read. To do that, I needed to understand  how all the use cases related to each other, and specifically whether those last two use cases had anything to do with permissions. I misinterpreted the previous answer to that question as "yes, these are all about permissions," which is how I ended up with that list.

I'd still like to understand what WAR and Externally Connectable are and how they use match patterns, but I don't need that information to fix the list.

Comment (AmySteam) [reply site/en/docs/extensions/mv3/match_patterns/index.md]: **Match pattern use explanation**

- The match patterns under web accessible resources declare which URLs are allowed to access specific resources that are bundled with the Chrome extension.

- The match patterns under "externally_connectable" declare which URLS are allowed to send and receive messages from the Chrome extension.

File: site/en/docs/extensions/mv3/match_patterns/index.md
Comment (wizardofdocs) [reply site/en/docs/extensions/mv3/match_patterns/index.md]: I think I've fixed it. Anything else I need to add?

File: site/en/docs/extensions/mv3/match_patterns/index.md
Comment (wizardofdocs) [reply site/en/docs/extensions/mv3/match_patterns/index.md]: So is it useful to say, for example, "Granting access to [web-accessible resources]" and "Allowing message sending and receiving using [`"externally_connectable.matches"`]" in the list?

Comment (AmySteam) [reply site/en/docs/extensions/mv3/match_patterns/index.md]: > So is it useful to say, for example, "Granting access to [web-accessible resources]" and "Allowing message sending and receiving using [`"externally_connectable.matches"`]" in the list?

This sounds like a good idea 👍.

Comment (wizardofdocs) [reply site/en/docs/extensions/mv3/match_patterns/index.md]: > > So is it useful to say, for example, "Granting access to [web-accessible resources]" and "Allowing message sending and receiving using [`"externally_connectable.matches"`]" in the list?
> 
> This sounds like a good idea 👍.

Done in the new commit.

File: site/en/docs/extensions/mv3/match_patterns/index.md
Comment (jpmedley) [reply site/en/docs/extensions/mv3/match_patterns/index.md]: @AmySteam

File: site/en/docs/extensions/mv3/match_patterns/index.md
Comment (jpmedley) [reply site/en/docs/extensions/mv3/match_patterns/index.md:21]: @AmySteam

Comment (AmySteam) [reply site/en/docs/extensions/mv3/match_patterns/index.md:21]: The  `/foo*` example below is for the **_path_**, not the **_host_**. So this is correct.

File: site/en/docs/extensions/mv3/match_patterns/index.md
Comment (jpmedley) [reply site/en/docs/extensions/mv3/match_patterns/index.md:30]: +1

File: site/en/docs/extensions/mv3/match_patterns/index.md
Comment (AmySteam) [inline site/en/docs/extensions/mv3/match_patterns/index.md:79]:
```suggestion
[permissions]: /docs/extensions/mv3/declare_permissions/#allow_access
[host-permissions]: /docs/extensions/mv3/declare_permissions/#host-permissions
```
Fix broken link

File: site/en/docs/extensions/mv3/match_patterns/index.md
Comment (AmySteam) [reply site/en/docs/extensions/mv3/match_patterns/index.md]: > '/' means the same thing as '/*', but do we really need to tell people that? It's what linguist Bryan Garner would call a [needless variant](https://www.cjr.org/analysis/there_ought_to_be_alternative.php#:~:text=might%20call%20a%20%E2%80%9C-,needless%20variant,-%E2%80%9D%3A%20%E2%80%9Ctwo%20or%20more)

It's just more explicit for developers to know if you add `*.example.com/`, it will also match `*.example.com/foo` and `*.example.com/bar`, not just `www.example.com`.

I think when it comes to match patterns, it's okay to be explicit.  But at this point I'm ready to agree with anything ya'll decide.

**Comment by jpmedley** [comment]
@oliverdunk Are you done?

File: site/en/docs/extensions/mv3/match_patterns/index.md
Comment (wizardofdocs) [reply site/en/docs/extensions/mv3/match_patterns/index.md]: Is this phrasing better?

File: site/en/docs/extensions/mv3/match_patterns/index.md
Comment (AmySteam) [inline site/en/docs/extensions/mv3/match_patterns/index.md:23]: Love it!!! 💯 🤩

File: site/en/docs/extensions/mv3/match_patterns/index.md
Comment (AmySteam) [reply site/en/docs/extensions/mv3/match_patterns/index.md]: image

**Review by jpmedley — APPROVED** [review]

**Comment by jpmedley** [comment]
@oliverdunk, are you done?

File: site/en/docs/extensions/mv3/match_patterns/index.md
Comment (oliverdunk) [inline site/en/docs/extensions/mv3/match_patterns/index.md:16]: Are we sure about `urn`?

File: site/en/docs/extensions/mv3/match_patterns/index.md
Comment (oliverdunk) [inline site/en/docs/extensions/mv3/match_patterns/index.md:29]:
```suggestion
* Granting access to [web-accessible resources][war].
```

File: site/en/docs/extensions/mv3/match_patterns/index.md
Comment (oliverdunk) [inline site/en/docs/extensions/mv3/match_patterns/index.md:44]: What does this mean? That you can't have `*.co.uk`?

File: site/en/docs/extensions/mv3/match_patterns/index.md
Comment (oliverdunk) [inline site/en/docs/extensions/mv3/match_patterns/index.md:70]: We should drop this if it doesn't work (it doesn't as far as I can tell from testing).

File: site/en/docs/extensions/mv3/match_patterns/index.md
Comment (oliverdunk) [inline site/en/docs/extensions/mv3/match_patterns/index.md:41]:
```suggestion
: To match any localhost port during development, use `http://localhost:*/*`. Specify a port using the port number, as in `http://localhost:8080/*`.  For IP addresses, specify the address plus a wildcard in the path, as in `http://127.0.0.1/*`. You can also use `http://*:*/*` to match localhost, IP addresses, and any port.
```

**Review by oliverdunk — CHANGES_REQUESTED** [review]
Left a few final comments. Thanks for the continued work on this one.

File: site/en/docs/extensions/mv3/match_patterns/index.md
Comment (jpmedley) [reply site/en/docs/extensions/mv3/match_patterns/index.md:16]: I'm trying to find out.

Comment (jpmedley) [reply site/en/docs/extensions/mv3/match_patterns/index.md:16]: If I understood Devlin correctly, this is the correct list of schemes. https://source.chromium.org/chromium/chromium/src/+/main:extensions/common/url_pattern.cc;l=31-52;drc=8c08dd32bf007ca09f57b63b6b812478b44ac2c4

**Comment by 0569395312** [comment]
SA‎8880000203608010980048

0569395312

في الأربعاء، ٢٠ سبتمبر ٢٠٢٣, ٧:٢٣ م Joe Medley ***@***.***>
كتب:

**Comment by 0569395312** [comment]
https://drive.google.com/file/d/1pV-qenYzqnmr3RV_t1BeMzDHzmBT4CKy/view?usp=drivesdk

0569395312

في الأربعاء، ٢٠ سبتمبر ٢٠٢٣, ٧:٢٤ م عثمان شافعي ***@***.***>
كتب:

File: site/en/docs/extensions/mv3/match_patterns/index.md
Comment (wizardofdocs) [reply site/en/docs/extensions/mv3/match_patterns/index.md:44]: The original says you can't do `google.*`. `*.example.com` is allowed.

File: site/en/docs/extensions/mv3/match_patterns/index.md
Comment (wizardofdocs) [reply site/en/docs/extensions/mv3/match_patterns/index.md]: Updated. Does it still need the single quotes?

File: site/en/docs/extensions/mv3/match_patterns/index.md
Comment (wizardofdocs) [reply site/en/docs/extensions/mv3/match_patterns/index.md]: Removed.

File: site/en/docs/extensions/mv3/match_patterns/index.md
Comment (jpmedley) [reply site/en/docs/extensions/mv3/match_patterns/index.md]: I'd say no quotes. I insist on quotes for manifest keys because they're part of what actually needs to be in the manifest. Quotes aren't part of a URL scheme.

Comment (wizardofdocs) [reply site/en/docs/extensions/mv3/match_patterns/index.md]: Done.

File: site/en/docs/extensions/mv3/match_patterns/index.md
Comment (jpmedley) [reply site/en/docs/extensions/mv3/match_patterns/index.md:57]: It would need to be `*://*.co.uk`, which is [in one of our unit tests](https://source.chromium.org/chromium/chromium/src/+/main:extensions/common/url_pattern_unittest.cc;l=1056;drc=707b3abc38ded8ab7eb0fa21664b1556324929b4). I'm told that gets the same result as `<all_urls>`. So, should we care?

File: site/en/docs/extensions/mv3/match_patterns/index.md
Comment (AmySteam) [inline site/en/docs/extensions/mv3/match_patterns/index.md]: I'm a bit uncomfortable adding all these scheme's because I know that extensions have restrictions. For example, what is the 'extension' scheme? If it's `chrome-extension://` URLs, these are not valid for content script injection.

Comment (jpmedley) [reply site/en/docs/extensions/mv3/match_patterns/index.md]: This list came from source code: https://source.chromium.org/chromium/chromium/src/+/main:extensions/common/url_pattern.cc;l=31-52;drc=8c08dd32bf007ca09f57b63b6b812478b44ac2c4

You're probably right about extension UI. (And a few other probably need to be spelled differently.) But confirm that in code yet.

File: site/en/docs/extensions/mv3/match_patterns/index.md
Comment (jpmedley) [reply site/en/docs/extensions/mv3/match_patterns/index.md:57]: @oliverdunk

File: site/en/docs/extensions/mv3/match_patterns/index.md
Comment (oliverdunk) [reply site/en/docs/extensions/mv3/match_patterns/index.md]: Notes:

- Chrome no longer supports FTP so I don't think `ftp` makes sense.
- I'm not familiar with (and couldn't find) any use case for `chromeui`. It seems internal so I would remove.
- The same for `extension` and `filesystem`.
- Looking at the code, we don't seem to actually support `data` in any APIs, so I would remove that.
- `uuid_in_package` was removed here: https://groups.google.com/a/chromium.org/g/blink-dev/c/iyhes6tMz2Y

So I think the full list should be:

```
http
https
A wildcard *, which matches only http or https
file
```

File: site/en/docs/extensions/mv3/match_patterns/index.md
Comment (AmySteam) [reply site/en/docs/extensions/mv3/match_patterns/index.md:57]: The TLDs section was ported directly from the extension's FAQs article. This may have been so in an older version of Chrome, since @oliverdunk has been able to test that content scripts can be injected into TLDs match patterns.

**Comment by Mohammed71o** [comment]
T ou criss, là, vous vous dite développer,non pirate

Le mar. 10 oct. 2023, 12 h 37, amysteamdev ***@***.***> a
écrit :

**Comment by Mohammed71o** [comment]
Ta rive tu je reste rais pas dehors toute la journée

**Comment by Mohammed71o** [comment]
Désolé madame mais il y a quelqu'un de githud qui dit venir ici

File: site/en/docs/extensions/mv3/match_patterns/index.md
Comment (jpmedley) [reply site/en/docs/extensions/mv3/match_patterns/index.md:57]: @AmySteam, I didn't follow that.

File: site/en/docs/extensions/mv3/match_patterns/index.md
Comment (jpmedley) [reply site/en/docs/extensions/mv3/match_patterns/index.md]: +1

File: site/en/docs/extensions/mv3/match_patterns/index.md
Comment (jpmedley) [reply site/en/docs/extensions/mv3/match_patterns/index.md:57]: @oliverdunk told me over chat that the current text is fine.l

**Comment by jpmedley** [comment]
@oliverdunk I can't merge this without your LGTM.

File: site/en/docs/extensions/mv3/match_patterns/index.md
Comment (AmySteam) [inline site/en/docs/extensions/mv3/match_patterns/index.md:29]:
```suggestion
- `http`
- `https`
- A wildcard `*`, which matches only `http` or `https`
- `file`
```
@wizardofdocs @jpmedley 
Could we apply this change before merging? [Here's the conversation where we agreed to simplify this list](https://github.com/GoogleChrome/developer.chrome.com/pull/7215#discussion_r1352008813)

**Review by AmySteam — CHANGES_REQUESTED** [review]
One request change before we merge please. Based on our team discussion:

https://github.com/GoogleChrome/developer.chrome.com/pull/7215#discussion_r1352008813

**Review by AmySteam — APPROVED** [review]
Thanks! 👍

**Review by oliverdunk — APPROVED** [review]
LGTM, thanks for the patience on some of the trickier bits here 👍

**Pull request merged**
